doc/async_model.md

```diff
@@ -0,0 +1,118 @@
+# Asynchronous model {#async_model}
+
+@tableofcontents
+
+Open62541++ adapts the the well-proven [asynchronous model of (Boost) Asio](https://think-async.com/asio/asio-1.28.0/doc/asio/overview/model.html). A key goal of Asio's asynchronous model is to support multiple composition mechanisms. This is achieved via completion tokens.
+
+## Completion tokens
+
+Open62541++ accepts completion tokens as the final argument of asynchronous operations.
+
+```cpp
+// Function signature of the completion handler: void(opcua::StatusCode, opcua::Variant&)
+template <typename CompletionToken = opcua::DefaultCompletionToken>
+auto opcua::services::readValueAsync(
+    opcua::Client& client,
+    const opcua::NodeId& id,
+    CompletionToken&& token = opcua::DefaultCompletionToken()
+);
+```
+
+Following completion tokens can be used (and described in more detail below):
+
+- A callback function (object) with the expected completion signature
+- `opcua::useFuture` to return a future object `std::future<T>`
+- `opcua::useDeferred` to return a callable for deferred execution
+- `opcua::useDetached` to detach the asynchronous operation
+- Custom tokens by providing template specializations for `opcua::AsyncResult`
+
+### Callback function
+
+If the user passes a function object as the completion token, the asynchronous operation behaves as previously described: the operation begins, and when the operation completes the result is passed to the callback. The callback function must match the expected signature:
+
+```cpp
+void(opcua::StatusCode code)              // for void result types
+void(opcua::StatusCode code, T& result);  // for non-void result types
+void(opcua::StatusCode code, T result);   // for non-void result types (trivially-copyable)
+```
+
+```cpp
+opcua::services::readValueAsync(
+    client,
+    id,
+    [](opcua::StatusCode code, opcua::Variant& result) {
+        // ...
+    }
+);
+```
+
+The status code `opcua::StatusCode` needs to be checked to determine the validity of the result.
+The callback is executed within the client's or server's event loop. Please make sure not to block the event loop with synchronous requests.
```

Comment:
It looks like you have gained some experience already with this case, could you explain what is happening in this scenario?

Looking at the code in open62541 I thought it should work to run the event loop in both threads, the one using the sync API and the one running only the event loop. Though it would not be clear in which thread the completion handler is executed.

Reply:
You are right, sync calls should not be a problem
But initiating a async call inside a completion handler and waiting for it would block the event loop, correct?

Reply:
Yes, in that case the user would need to take care to do a context switch to another thread first.

Reply:
Just fixed the description in the documentation.